R/read-gs.R

```diff
@@ -15,11 +16,200 @@ read_gs <- function(path, verbose = FALSE){
+        " data due to value of source parameter. ",
+        "However, the data appears to be from source ",
+        auto_determine, "."
+      ))
+    }
+    #the read_canvas_grades does all of the lifting
+    return(read_canvas_grades(grades))
+    
+  } else {
+    stop(paste0(
+      "Specified source type ", source,
+      " is not currently supported."
+    ))
+  }
+  
+  
+ 
+}
+
+#'Convert Canvas grade dataframe into standard format
+#'
+#'Takes in a dataframe uploaded from Canvas. 
+#'Returns a dataframe in our standardized format (ie like Gradescope).
+#'
+#'
+#'@importFrom stringr str_extract str_match
+#'@importFrom dplyr filter select slice rename_with rename left_join
+#'@importFrom purrr discard
+#'@keywords internal
+read_canvas_grades <- function(grades){
+  #first determine which columns are assignments
+  
+  #canvas categories will not be kept
+  
+  assignments <- stringr::str_extract(names(grades),
+                                      ".+\\s*\\(\\d+\\)") |>
+                purrr::discard(is.na)
+  
+ 
+  #now reconfigure the points possible info
+  max_points <- dplyr::filter(grades, Student == "    Points Possible") |>
+    dplyr::select(all_of(assignments)) |>
+    dplyr::slice(rep.int(1, nrow(grades))) |>
+    dplyr::rename_with(function(x){
+      paste0(x, " - Max Points")
+    })
+  #add ID column to max points to facilitate matching
+  max_points$ID <- grades$ID
+  
+  #remove unneeded cols, add max_point column, and remove rows that are not students
+  grades <- grades |>
+    dplyr::select(c(all_of(assignments), "ID", "Student", 
+                    "SIS User ID", "Section")) |>
+    dplyr::left_join(max_points, 
+                     by = dplyr::join_by(ID == ID)) |>
+    dplyr::filter(!(Student %in% c("    Points Possible", "Student, Test")))
+  
+  #add columns for submission time and lateness
+  
+  sub_cols <- paste0(assignments, " - Submission Time")
+  
+  late_cols <- paste0(assignments, " - Lateness (H:M:S)")
+  
+  grades[sub_cols] <- as.POSIXct(NA)
+  
+  grades[late_cols] <- NA
+  
+  grades$`First Name` <- (stringr::str_match(grades$Student, 
+                                             ".+,\\s(.+)"))[, 2]
+  
+  grades$`Last Name` <- (stringr::str_match(grades$Student, 
+                                            "(.+),.+"))[, 2]
+  
+  
+  grades <- grades |>
+    dplyr::rename(SID = `SIS User ID`,
+                  Sections = Section) |>
+    dplyr::select(c("First Name", "Last Name", "SID", "Sections",
+                    as.vector(rbind(assignments, 
+                                    paste0(assignments, " - Max Points"), 
+                                    sub_cols,
+                                    late_cols)
+                              )
+                    )
+                  )
+  
+  attr(grades, "source") <- "Canvas"
+  
+  check_data_format(grades)
+  
 }
 
-#' Check Formatting of Gradescope Data
+#'Predict the source of grade dataframe.
```

Comment:
This function looks good! Something we'll want to keep an eye out for in the future: how we'll version `"Canvas"` and `"Gradescope"` as their schema (and presumably col names) change.